Fix: AR subclasses to work with AC::Parameters

claudiob opened the pull request:
Fixes #21407

The new test fails before this commit and passes now.
Adding `to_h` was suggested by @sikachu in Basecamp.

What is your opinion about the `require 'action_controller/railtie'` I had to add?

In principle, AR should not depend on AC, but this _principle_ is already ignored in [activerecord/railtie](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/railtie.rb#L5-L9).

Let me know :ribbon:

claudiob committed:
Fix: AR subclasses should work with AC::Parameters

Fixes #21407

The new test fails before this commit and passes now.
Adding `to_h` was suggested by @sikachu in Basecamp.

schneems commented:
r? @sikachu :wink:

sikachu commented:
Travis didn't like it.

> What is your opinion about the require 'action_controller/railtie' I had to add?

Can we require only `action_controller/metal/strong_parameters`?

claudiob committed:
Add missing require to strong_parameters.rb

The file [references Rack::Test here](https://github.com/rails/rails/blame/master/actionpack/lib/action_controller/metal/strong_parameters.rb#L671)
so it's better off requiring 'rack/test' in the first place.

egilburg commented on activerecord/lib/active_record/inheritance.rb:
```diff
@@ -192,11 +192,11 @@ def type_condition(table = arel_table)
       # If this is a StrongParameters hash, and access to inheritance_column is not permitted,
```
Is this comment still applicable to new behavior?

claudiob replied:
Mmmh… nice catch! This is what would happen:

```ruby
def test_new_with_unpermitted_action_controller_parameters
  params = ActionController::Parameters.new(type: 'Firm') # note: there is no permit here
  firm = Company.new params
  assert_equal Company, firm.class
end
```

results in

```
ActiveModel::ForbiddenAttributesError: ActiveModel::ForbiddenAttributesError
    /Users/claudiob/code/rails/activemodel/lib/active_model/forbidden_attributes_protection.rb:21:in `sanitize_for_mass_assignment'
    /Users/claudiob/code/rails/activemodel/lib/active_model/attribute_assignment.rb:33:in `assign_attributes'
    /Users/claudiob/code/rails/activerecord/lib/active_record/core.rb:305:in `initialize'
    /Users/claudiob/code/rails/activerecord/lib/active_record/inheritance.rb:61:in `new'
    /Users/claudiob/code/rails/activerecord/lib/active_record/inheritance.rb:61:in `new'
```

claudiob replied:
@sikachu Because of the above, adding `to_h` might not be enough… what do you think?

claudiob replied:
Actually… on a second thought I think it's correct that `ActiveModel::ForbiddenAttributesError` gets raised in this case. So my suggestion would be to simply remove the comment that @egilburg pointed to.

claudiob replied:
And to add a line to the CHANGELOG.md of ActiveRecord explaining that the behavior has changed: passing Parameters with an unpermitted `type` field will now raise an error (rather than being silently ignored)

claudiob committed:
Remove old comment about AC::Parameters>subclasses

[ci skip]

Q: What happens if you initialize an AR model by passing Parameters that
have not been whitelisted with `permit`?

A: An `ActiveModel::ForbiddenAttributesError` is raised.

I think this behavior is correct, and it's better than what used to happen,
with unpermitted parameter being simply ignored.

claudiob commented:
@egilburg @sikachu See my latest commit for the comment part.

Also that change should be included in the Rails 5 release notes, so I'm referencing https://github.com/rails/rails/issues/20612 here :ribbon:

claudiob commented:
Now that #17169 has been merged into master, the core commit of this PR does not need to be anymore.

However, the other two commits that change documentation, test and require still make sense, so I'm extracting them into separate PRs.
